### DISCUSSION TITLE
Problems accessing `scipp.constants`

### DISCUSSION THREAD

**Discussion by arm61** [scipp/scipp#3561] [Scipp Q&A] [answered]
The following raises an error for me: 

```
import scipp as sc
sc.constants
```

But works if I do

```
from scipp import constants
```

Scipp version: 24.9.1

**Comment by SimonHeybrock** [comment, marked as answer]
Hi! This is by design. Similar to, say SciPy submodules, it is a submodule that needs to be imported explicitly. One reason for this is that it relies on Scipy, which is an optional dependency of Scipp.

We will consider adding lazy loading, please see #3567.

For now, if you would like to use `sc.constants` in code, you can

```python
import scipp as sc
import scipp.constants

sc.constants.h  # works now
```

**Reply by arm61** [reply]
```
import scipy as sp
sp.constants
```

Doesn't raise an error.

**Reply by arm61** [reply]
The scipy dependency argument is fair, though.

**Reply by SimonHeybrock** [reply]
Hmm, at least some of Scipy's submodules used to do that.

**Reply by arm61** [reply]
Perhaps the `physical_constants` subsubmodule?

**Reply by SimonHeybrock** [reply]
Seemed to be all: https://github.com/scipy/scipy/blob/4d7413563cfa477c08717daa61e5994e968ee79f/scipy/__init__.py#LL15C1-L16C26

**Reply by jl-wynen** [reply]
Scipy imports modules lazily when you access them: https://github.com/scipy/scipy/blob/1adc8cb7a37e09444632a86afea5ee1d07f3e3c5/scipy/__init__.py#L134
We could do the same but I am not sure it is worth the complication.

**Reply by SimonHeybrock** [reply]
There is actually a SPEC about it: https://scientific-python.org/specs/spec-0001/

**Reply by arm61** [reply]
Does a SPEC mean that you will change in the future?

**Reply by SimonHeybrock** [reply]
It is something the "Scientific Python" community agreed on. So there is nothing forcing us to do so, but it would make a lot of sense. Please subscribe to #3567 for more updates on the topic.
